File: Data Structures & Algorithms/minimum-size-subarray-sum/submission-2.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int n = nums.size();
        int left = 0, right = 0;
        int sum = 0, res = n+1;

        while(right < n) {
            sum += nums[right];

            while(sum >= target) {
                sum -= nums[left];
                res = min(res, right - left + 1);
                left++;
            }

            // res = min(res, right - left + 1);
            right++;
        }
        res = min(res, right-left+1);

        return res == n+1 ? 0 : res;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        res = []

        for i in range(0, n-2):
            l = i+1
            r = n-1

            if i > 0 and nums[i] == nums[i-1]:
                continue
            req = -nums[i]

            while l < r:
                if nums[l] + nums[r] == req:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    r -= 1

                    while l < r and nums[l] == nums[l - 1]:
                        l += 1

                    while l < r and nums[r] == nums[r + 1]:
                        r -= 1
                
                elif nums[l] + nums[r] > req:
                    r -= 1
                
                else: 
                    l += 1
        
        return res


